pkg/operator/apis/monitoring/v1alpha1/types.go

```diff
@@ -219,20 +297,80 @@ func (pm *PodMonitoring) ScrapeConfigs() (res []*promconfig.ScrapeConfig, err er
 	return res, nil
 }
 
+// SetPodMonitoringCondition merges the provided PodMonitoring resource to the
+//
+// along with the provided condition iff the resource generation has changed or there
+// is a status condition state transition.
+func (status *PodMonitoringStatus) SetPodMonitoringCondition(gen int64, now metav1.Time, cond *MonitoringCondition) (bool, error) {
+	var (
+		specChanged              = status.ObservedGeneration != gen
+		statusTransition, update bool
+		conds                    = make(map[MonitoringConditionType]*MonitoringCondition)
+	)
+
+	if cond.Type == "" || cond.Status == "" {
+		return update, errInvalidCond
+	}
+
+	// Set up defaults.
+	for _, mc := range NewDefaultConditions(now) {
+		conds[mc.Type] = &mc
+	}
+	// Overwrite with any previous state.
+	for _, mc := range status.Conditions {
+		conds[mc.Type] = &mc
+	}
+
+	// Set some timestamp defaults if unspecified.
+	cond.LastUpdateTime = now
+
+	// Check if the condition results in a transition of status state.
+	if old := conds[cond.Type]; old.Status == cond.Status {
+		cond.LastTransitionTime = old.LastTransitionTime
+	} else {
+		cond.LastTransitionTime = cond.LastUpdateTime
+		statusTransition = true
+	}
+
+	// Set condition.
+	conds[cond.Type] = cond
+
+	// Only update status if the spec has changed (indicated by Generation field) or
+	// if this update transitions status state.
+	if specChanged || statusTransition {
+		update = true
+		status.ObservedGeneration = gen
+		status.Conditions = status.Conditions[:0]
+		for _, c := range conds {
+			status.Conditions = append(status.Conditions, *c)
+		}
+	}
+
+	return update, nil
+}
+
 // Environment variable for the current node that needs to be interpolated in generated
 // scrape configurations for a PodMonitoring resource.
 const EnvVarNodeName = "NODE_NAME"
 
-func (pm *PodMonitoring) endpontScrapeConfig(index int) (*promconfig.ScrapeConfig, error) {
+type monitoringScrapeConfig interface {
+	GetNamespace() string
+	GetName() string
+	GetSelector() metav1.LabelSelector
+	GetTargetLabels() TargetLabels
+}
+
+func endpointScrapeConfig(msc monitoringScrapeConfig, ep ScrapeEndpoint) (*promconfig.ScrapeConfig, error) {
+	var (
+		ns           = msc.GetNamespace()
+		name         = msc.GetName()
+		sel          = msc.GetSelector()
+		targetLabels = msc.GetTargetLabels()
+	)
```

Comment:
Why not just add these 4 as args directly rather than defining an interface and adding methods to both types?
The signature sure is longer but it's overall a good chunk less stuff to comprehend. Also makes refactoring  simpler when we add differentiating attributes and this single generation function for both CRDs will no longer suffice.